### DISCUSSION TITLE
Add feature for useParse in myst-to-react to render mdast output Node

### DISCUSSION THREAD

**Discussion by GYHHAHA** [executablebooks/meta#987] [Ideas] [answered]
Currently, it seems that the content generated from useParse doesn't include the output cell. It's a good idea to implement the output node render feature which works like reduceOutputs in myst-cli. Or do we can any alternative for the output render? @rowanc1 Thanks!
https://github.com/executablebooks/myst-theme/blob/bd8556fc7f63e1933b01e5125e74ff2c503ed031/packages/myst-to-react/src/index.tsx#L48
https://github.com/executablebooks/mystjs/blob/bc466578af3c54341d4008755bb9d82f5ad28672/packages/myst-cli/src/transforms/outputs.ts#L66

**Comment by rowanc1** [comment, marked as answer]
The output renderer is brought in by the `@myst-theme/jupyter` package ([here](https://github.com/executablebooks/myst-theme/blob/bd8556fc7f63e1933b01e5125e74ff2c503ed031/packages/jupyter/src/index.tsx#L3)). We split them up as there are a lot of conflicts in packages when using this inside of `jupyterlab-myst` (i.e. we want to use jupyter natively in that case, not bring in a subset of the capabilities). The output renderer can be brought in directly if you are doing your own renderer using something like [this](https://github.com/executablebooks/myst-theme/blob/bd8556fc7f63e1933b01e5125e74ff2c503ed031/packages/site/src/components/renderers.ts#L10) and then use the theme-provider ([here](https://github.com/executablebooks/myst-theme/blob/bd8556fc7f63e1933b01e5125e74ff2c503ed031/packages/site/src/pages/Root.tsx#L49)) or the `useParse` function directly to give you access to those renderers in react!

**Reply by GYHHAHA** [reply]
Thanks! I will take a look.

**Reply by rowanc1** [reply]
I would also note that @stevejpurves is doing a bunch of work at the moment to bring thebe (using jupyter and jupyter-lite) into the theme, and so there might be quite a few updates there shortly. :)

For example, this is attaching binder to a myst rendered page -- super excited about this!!

https://user-images.githubusercontent.com/913249/228981938-a78cef27-159e-4f33-be5f-89198b0f8a36.mp4

**Reply by GYHHAHA** [reply]
Great! This looks pretty good. I will dive into this after addressing the useParse problem.

**Reply by GYHHAHA** [reply]
I'm pretty exciting that this solution finally works! I successfully replace the original _static cache folder with AWS S3 by customizing `transformOutputs` function. Thanks @rowanc1

**Reply by rowanc1** [reply]
Awesome, glad that is working for you! There is also a function [here](https://github.com/executablebooks/myst-theme/blob/main/packages/site/src/loaders/utils.ts#L109) that allows you to traverse a single page and dynamically update the links based on a CDN. I think that needs to likely be over in the mystjs repo in `myst-common`. If you look at the [cdn-loader](https://github.com/executablebooks/myst-theme/blob/main/packages/site/src/loaders/cdn.server.ts#L142) there are some other hints at how to dynamically load from a CDN (i.e. look up domain hostname --> CDN id, then load the content from a known CDN/bucket).

If your solution is even somewhat sharable, I would love to see your approach. In the next month or so we need to write up documentation for deployment. Deploying to an AWS bucket would be an awesome target!

**Reply by GYHHAHA** [reply]
Great, I will look them soon! However there is still one single step I'm still not satisfied: Since I need to `import { processNotebook } from "myst-cli";` and I haven't find a correct way to do this inside the remix.js, which means I have to launch a separate child process to execute the node program on the compiled file from esbuild like your playground repo. Hope there is a nicer way in the future to do that!

**Reply by rowanc1** [reply]
Hm, generally the CLI isn't designed to be imported into remix, certainly not in the frontend code as it has file-system dependencies (and I think it might break the backend stuff as well). A quick check that *might* work, rename a utility to `utils.server.ts` (see [remix docs](https://remix.run/docs/en/1.15.0/pages/gotchas#server-code-in-client-bundles)) and do the processing in there, that forces their bundler to exclude it from client code. If that doesn't work, then we should aim to move the utility to something that can be imported in more places (e.g. myst-transforms).

**Reply by GYHHAHA** [reply]
Thanks, I will take a try for that.

**Reply by GYHHAHA** [reply]
The strategy that change file name to `*.server.ts` works perfectly. Thanks for let me know! @rowanc1 Now here is a quick demo for the jupyter output render.

https://user-images.githubusercontent.com/41546976/231354909-b91077ca-187b-456f-9950-7eb892e698d7.mp4

I have another question: since the execution for notebook book is pretty slow, suppose I got 10 notebooks to render. After I click the start button, I want to know from the client side that how many notebooks are finished at the server side, should I use websocket tools like socket.io? Is there any better solution?
